### DISCUSSION TITLE
Can't use array functions on arrays stored in globalChannelMap

### DISCUSSION THREAD

**Discussion by bartlettmichael62** [nextgenhealthcare/connect#4542] [Q&A] [answered]
We're trying to store an array of Strings (one-dimensional array) from the Deploy Script to the globalChannelMap, but a code template we're invoking in the Destination Filter isn't able to recognize the object we're pulling from globalChannelMap as an array. We can still pull items from it via index (i.e. $('globalArray')[0] returns the proper value), but we can't use functions like **indexOf()** as shorthand to check if an element exists in the array.

### Deploy Script:

Assume 'mappedValues' is this String: "F,C,A,P,Verified,Auth (Verified)" (it's stored off-server). We're splitting it into an array, using the comma (',') as a delimiter:
`labStatusList = mappedValues.split(',');` //Returns [F, C, A, P, Verified, Auth (Verified)]
`globalChannelMap.put('acceptedLabStatus', labStatusList);`

During troubleshooting, I put the following code in that Code Template to try and figure out what Mirth is doing. **var localArray** is just a copy of the same array we stored in the globalChannelMap back in the Deploy Script, just locally scoped to the Code Template that's giving us issues:

### Destination Filter Code Template:

`var localArray = ['F', 'C', 'A', 'P', 'Verified', 'Auth (Verified)'];`
`var copiedArray = $('acceptedLabStatus');`        //Stored on globalChannelMap ahead of time

//Assume var labStatus = 'F' (String)
`channelMap.put('_debug1', localArray.indexOf(labStatus) );`
`channelMap.put('_debug2', copiedArray.indexOf(labStatus) );`

debug1 => 0.0 (first element)
debug2 => -1.0 (not found)

We're just trying to use the array from globalChannelMap to perform this type of check on input values:

`if ($('acceptedLabStatus').indexOf(inputLabStatus) == -1) {`
`        channelMap.put(filterNote, "blah blah blah");`
`        return false;`
`}`

I've tried seemingly everything to re-cast the object from the globalChannelMap as another type and nothing has worked. We can still reference the global as an array if we specify an index. $('acceptedLabStatus')[0] still returns the first element of our array, as we'd expect. The problem is that we're locked out of all the special functions that come with Arrays. I DO NOT want to have to explicitly write out a loop to iterate through each element of $('acceptedLabStatus') and check every element, because that's ridiculous and we shouldn't have to do it; functions like 'indexOf()' exist for a reason. There's multiple spots in our code where we need to store an array of mapped values globally, and then check whether or not an inbound values exists in those arrays.  You're just going to have to trust me when I say that the extra loops would introduce a lot of extra jank to our code. Sure, it would be functionally equivalent, it just looks really bad.

Object.prototype.toString.call(**localArray**) ---> [object Array]
Object.prototype.toString.call(**copiedArray**) ---> [object JavaArray]

### Main Question:

How could we recast the array from globalChannelMap to get it to cooperate with the standard suite of Array functions? Specifically something like **indexOf()**. Alternatively, could somebody point me to the nearest Mirth documentation for [object JavaArray] and the list of functions it supports?

Thank You!!

**Comment by bartlettmichael62** [comment, marked as answer]
I slept and figured it out:

**.split()** was returning a JavaArray, so I'm iterating an extra time to shove all the elements into a native (Javascript) array before pushing it to the globalChannelMap.

`labStatusList = [];`
`tempArray = mappedValues.split(',');`
`for each (var item in tempArray) {`
    `labStatusList.push(item);`
`}`

**Comment by tonygermano** [comment]
You actually **can** call all Array.prototype methods on java arrays. Having the wrong value returned by indexOf instead of throwing an exception that the method was not found meant that it was working.

The fact that split produced a java array means that you started with a java String, and your resulting array, therefore, contained java Strings because it called the Java version of `String.split`. Calling indexOf on this array of java Strings and passing it a javascript string to check meant that indexOf couldn't find a match because it uses strict equality and java Strings are objects, so will not match to a string.

Changing the first line of your original code should fix it all. Converting mappedValues to a javascript string, means that split will give you a javascript array of javascript strings.
```js
labStatusList = String(mappedValues).split(',');
```

**Reply by tonygermano** [reply]
You could also throw that into a Java Set, which would then give you a contains method that would work on either java or javascript strings.
```js
// Convert mappedValues to javascript string, then split, then pass to new HashSet constructor.
// Javascript arrays implement the Java Collection and List interfaces when passed to Java
// methods that expect them.
var labStatusSet = new java.util.HashSet(String(mappedValues).split(','));
globalChannelMap.put('acceptedLabStatus', labStatusSet);
channelMap.put('_debug1', $('acceptedLabStatus').contains(labStatus));
channelMap.put('_debug2', $('acceptedLabStatus').contains(String(labStatus))); // explicitly check javascript string
channelMap.put('_debug3', $('acceptedLabStatus').contains(new java.lang.String(labStatus))); // explicitly check java String
```

**Reply by bartlettmichael62** [reply]
Hi Tony,

You're right on the money! As soon as I posted that this morning, I quickly found out that my solution only worked on an input of 'F'. As soon as the strings got larger than one character, the comparisons were all off. I was catching up on emails and read your first reply, and I figured I should probably mention the additional need to cast with **String()** from deployment. I fell into a trap using **toString()** for all my string-casting (not a word), without recognizing the crossover in Mirth between javascript and java.

Thanks for looking at these forums. It's good to know that someone's out there when you're at wits end.

Thanks again,
Mike
